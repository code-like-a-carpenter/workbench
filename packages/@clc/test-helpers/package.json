{
  "name": "@clc/test-helpers",
  "description": "test helpers",
  "private": true,
  "dependencies": {
    "@aws-sdk/lib-dynamodb": "3.726.1",
    "@clc/dependencies": "*"
  },
  "engines": {
    "node": "22.x"
  },
  "author": "Ian Remmel <1182361+ianwremmel@users.noreply.github.com> (https://www.ianwremmel.com)",
  "bugs": "https://www.github.com/code-like-a-carpenter/workbench/issues",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/types/index.d.mts",
        "default": "./dist/esm/index.mjs"
      },
      "require": {
        "types": "./dist/cjs-types/index.d.ts",
        "default": "./dist/cjs/index.cjs"
      }
    },
    "./package.json": "./package.json"
  },
  "homepage": "https://www.github.com/code-like-a-carpenter/workbench/tree/main/packages/@clc/test-helpers",
  "license": "MIT",
  "repository": "git@github.com:code-like-a-carpenter/workbench.git",
  "type": "module",
  "types": "./dist/types/index.d.mts",
  "publishConfig": {
    "access": "public"
  }
}
