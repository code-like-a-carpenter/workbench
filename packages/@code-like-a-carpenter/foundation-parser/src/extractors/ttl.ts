import type {GraphQLObjectType} from 'graphql';
import {isNonNullType} from 'graphql';

import {assert, fail} from '@code-like-a-carpenter/assert';
import type {TTLConfig} from '@code-like-a-carpenter/foundation-intermediate-representation';

import {
  getDirective,
  getOptionalArgBooleanValue,
  getOptionalArgStringValue,
} from '../helpers';

/**
 * Accepts a duration string and returns a number representing the duration in
 * milliseconds.
 */
export function convertDuration(duration: string): number {
  const durationUnit = duration.slice(-1);
  const durationValue = duration.slice(0, -1);

  assert(
    !Number.isNaN(durationValue),
    `Invalid ttl duration value: ${durationValue}`
  );

  switch (durationUnit) {
    case 's':
      return Number(durationValue) * 1000;
    case 'm':
      return Number(durationValue) * 1000 * 60;
    case 'h':
      return Number(durationValue) * 1000 * 60 * 60;
    case 'd':
      return Number(durationValue) * 1000 * 60 * 60 * 24;
    default:
      fail(`Invalid ttl duration: ${duration}. Unit must be one of s, m, h, d`);
  }
}

/**
 * Determines TTL configuration for a particular Model.
 */
export function extractTTLConfig(
  type: GraphQLObjectType
): TTLConfig | undefined {
  const fields =
    type.astNode?.fields?.filter((field) =>
      field.directives?.map(({name}) => name.value).includes('ttl')
    ) ?? [];
  if (fields.length === 0) {
    return undefined;
  }
  assert(fields.length === 1, 'Only one field can be marked with @ttl');
  const [astField] = fields;
  const fieldName = astField.name.value;

  const directive = getDirective('ttl', astField);

  const duration = getOptionalArgStringValue('duration', directive);

  if (!duration) {
    const field = type.getFields()[fieldName];

    return {
      argumentAllowed: true,
      argumentRequired: isNonNullType(field.type),
      fieldName,
    };
  }

  const seconds = convertDuration(duration);

  return {
    argumentAllowed: !!getOptionalArgBooleanValue('overridable', directive),
    argumentRequired: false,
    duration: seconds,
    fieldName,
  };
}
