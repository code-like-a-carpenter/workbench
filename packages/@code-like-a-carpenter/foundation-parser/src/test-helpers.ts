import {readFile} from 'node:fs/promises';
import path from 'node:path';

import {buildSchema} from 'graphql';

import {parse} from './parser';

export async function parseSchema(raw: string) {
  const coreSchema = await readFile(
    path.resolve(__dirname, '..', '..', '..', '..', './schema.graphqls'),
    'utf8'
  );

  const schema = buildSchema(`${coreSchema}\n${raw}`);

  return parse(
    schema,
    [],
    {
      defaultDispatcherConfig: {
        memorySize: 384,
        timeout: 90,
      },
      defaultHandlerConfig: {
        memorySize: 256,
        timeout: 30,
      },
      dependenciesModuleId: 'PLACEHOLDER',
    },
    {
      outputFile: './placeholder/output',
    }
  );
}
