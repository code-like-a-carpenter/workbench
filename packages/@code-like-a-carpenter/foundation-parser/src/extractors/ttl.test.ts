import {parseSchema} from '../test-helpers';

import {convertDuration} from './ttl';

describe('@ttl', () => {
  it('marks a fields as the ttl field', async () => {
    const schema = /* GraphQL */ `
      type Example implements Model & Timestamped & Versioned
        @partitionKey(pkFields: ["id"]) {
        createdAt: Date!
        expires: Date! @ttl
        id: ID!
        updatedAt: Date!
        version: Int!
      }
    `;

    const parsed = await parseSchema(schema);

    expect(parsed.tables[0].hasTtl).toBe(true);

    const [{ttlConfig}] = parsed.models;
    expect(ttlConfig).toBeDefined();
    expect(ttlConfig?.argumentAllowed).toBe(true);
    expect(ttlConfig?.argumentRequired).toBe(true);

    expect(parsed.models[0].ttlConfig).toMatchInlineSnapshot(`
      {
        "argumentAllowed": true,
        "argumentRequired": true,
        "fieldName": "expires",
      }
    `);
    expect(parsed).toMatchSnapshot();
  });

  it('expires a record after a specified duration', async () => {
    const schema = /* GraphQL */ `
      type Example implements Model & Timestamped & Versioned
        @partitionKey(pkFields: ["id"]) {
        createdAt: Date!
        expires: Date! @ttl(duration: "1d")
        id: ID!
        updatedAt: Date!
        version: Int!
      }
    `;

    const parsed = await parseSchema(schema);

    expect(parsed.tables[0].hasTtl).toBe(true);

    const [{ttlConfig}] = parsed.models;
    expect(ttlConfig).toBeDefined();
    expect(ttlConfig?.argumentAllowed).toBe(false);
    expect(ttlConfig?.argumentRequired).toBe(false);

    expect(parsed.models[0].ttlConfig).toMatchInlineSnapshot(`
      {
        "argumentAllowed": false,
        "argumentRequired": false,
        "duration": 86400000,
        "fieldName": "expires",
      }
    `);
    expect(parsed).toMatchSnapshot();
  });

  it('optionally allows the end-user to specify an alternative expiration', async () => {
    const schema = /* GraphQL */ `
      type Example implements Model & Timestamped & Versioned
        @partitionKey(pkFields: ["id"]) {
        createdAt: Date!
        expires: Date! @ttl(duration: "1d", overridable: true)
        id: ID!
        updatedAt: Date!
        version: Int!
      }
    `;

    const parsed = await parseSchema(schema);

    expect(parsed.tables[0].hasTtl).toBe(true);

    const [{ttlConfig}] = parsed.models;
    expect(ttlConfig).toBeDefined();

    expect(ttlConfig?.argumentAllowed).toBe(true);
    expect(ttlConfig?.argumentRequired).toBe(false);

    expect(parsed.models[0].ttlConfig).toMatchInlineSnapshot(`
      {
        "argumentAllowed": true,
        "argumentRequired": false,
        "duration": 86400000,
        "fieldName": "expires",
      }
    `);
    expect(parsed).toMatchSnapshot();
  });
});

describe('convertDuration()', () => {
  it('converts a duration string to milliseconds', () => {
    expect(convertDuration('1s')).toBe(1000);
    expect(convertDuration('1m')).toBe(60000);
    expect(convertDuration('1h')).toBe(3600000);
    expect(convertDuration('1d')).toBe(86400000);
  });
});
