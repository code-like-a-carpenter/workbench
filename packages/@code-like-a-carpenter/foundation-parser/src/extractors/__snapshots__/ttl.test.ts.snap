// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ttl expires a record after a specified duration 2`] = `
{
  "additionalImports": [],
  "dependenciesModuleId": "dependencies",
  "models": [
    {
      "consistent": false,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "ttl",
          "columnNamesForRead": [
            "ttl",
          ],
          "computeFunction": undefined,
          "ean": ":expires",
          "eav": "#expires",
          "fieldName": "expires",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
      ],
      "isLedger": false,
      "isPublic": false,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "enableEncryption": "IsProd",
        "enablePointInTimeRecovery": "IsProd",
        "hasPublicModels": false,
        "hasTtl": true,
        "primaryKey": {
          "isComposite": false,
          "isSingleField": true,
          "partitionKeyFields": [
            {
              "columnName": "id",
              "columnNamesForRead": [
                "id",
                "id",
              ],
              "computeFunction": undefined,
              "ean": ":id",
              "eav": "#id",
              "fieldName": "id",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "ID",
            },
          ],
          "partitionKeyPrefix": undefined,
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TableExample",
      },
      "ttlConfig": {
        "argumentAllowed": false,
        "argumentRequired": false,
        "duration": 86400000,
        "fieldName": "expires",
      },
      "typeName": "Example",
    },
  ],
  "tables": [
    {
      "enableEncryption": "IsProd",
      "enablePointInTimeRecovery": "IsProd",
      "hasPublicModels": false,
      "hasTtl": true,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TableExample",
    },
  ],
}
`;

exports[`@ttl marks a fields as the ttl field 2`] = `
{
  "additionalImports": [],
  "dependenciesModuleId": "dependencies",
  "models": [
    {
      "consistent": false,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "ttl",
          "columnNamesForRead": [
            "ttl",
          ],
          "computeFunction": undefined,
          "ean": ":expires",
          "eav": "#expires",
          "fieldName": "expires",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
      ],
      "isLedger": false,
      "isPublic": false,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "enableEncryption": "IsProd",
        "enablePointInTimeRecovery": "IsProd",
        "hasPublicModels": false,
        "hasTtl": true,
        "primaryKey": {
          "isComposite": false,
          "isSingleField": true,
          "partitionKeyFields": [
            {
              "columnName": "id",
              "columnNamesForRead": [
                "id",
                "id",
              ],
              "computeFunction": undefined,
              "ean": ":id",
              "eav": "#id",
              "fieldName": "id",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "ID",
            },
          ],
          "partitionKeyPrefix": undefined,
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TableExample",
      },
      "ttlConfig": {
        "argumentAllowed": true,
        "argumentRequired": true,
        "fieldName": "expires",
      },
      "typeName": "Example",
    },
  ],
  "tables": [
    {
      "enableEncryption": "IsProd",
      "enablePointInTimeRecovery": "IsProd",
      "hasPublicModels": false,
      "hasTtl": true,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TableExample",
    },
  ],
}
`;

exports[`@ttl optionally allows the end-user to specify an alternative expiration 2`] = `
{
  "additionalImports": [],
  "dependenciesModuleId": "dependencies",
  "models": [
    {
      "consistent": false,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "ttl",
          "columnNamesForRead": [
            "ttl",
          ],
          "computeFunction": undefined,
          "ean": ":expires",
          "eav": "#expires",
          "fieldName": "expires",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
      ],
      "isLedger": false,
      "isPublic": false,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "enableEncryption": "IsProd",
        "enablePointInTimeRecovery": "IsProd",
        "hasPublicModels": false,
        "hasTtl": true,
        "primaryKey": {
          "isComposite": false,
          "isSingleField": true,
          "partitionKeyFields": [
            {
              "columnName": "id",
              "columnNamesForRead": [
                "id",
                "id",
              ],
              "computeFunction": undefined,
              "ean": ":id",
              "eav": "#id",
              "fieldName": "id",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "ID",
            },
          ],
          "partitionKeyPrefix": undefined,
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TableExample",
      },
      "ttlConfig": {
        "argumentAllowed": true,
        "argumentRequired": false,
        "duration": 86400000,
        "fieldName": "expires",
      },
      "typeName": "Example",
    },
  ],
  "tables": [
    {
      "enableEncryption": "IsProd",
      "enablePointInTimeRecovery": "IsProd",
      "hasPublicModels": false,
      "hasTtl": true,
      "primaryKey": {
        "isComposite": false,
        "isSingleField": true,
        "partitionKeyFields": [
          {
            "columnName": "id",
            "columnNamesForRead": [
              "id",
              "id",
            ],
            "computeFunction": undefined,
            "ean": ":id",
            "eav": "#id",
            "fieldName": "id",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "ID",
          },
        ],
        "partitionKeyPrefix": undefined,
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TableExample",
    },
  ],
}
`;
